### PR TITLE
Add Missing MPI_F_XXX C constants

### PR DESCRIPTION
**Pull request by drwootton**
Add missing MPI_F_XXX C constants to mpi.h
This resolves issue #10391 

Signed-off-by: David Wootton <dwootton@us.ibm.com>

### CODE REVIEW

File: ompi/include/mpi.h.in [hunk 1]
```diff
@@ -571,6 +571,14 @@ typedef MPI_Win_errhandler_function MPI_Win_errhandler_fn
 #define MPI_MAX_PSET_NAME_LEN    OPAL_MAX_PSET_NAME_LEN /* max pset name len */
 #define MPI_MAX_STRINGTAG_LEN    OPAL_MAX_STRINGTAG_LEN /* max length of string arg to comm from group funcs*/
 
+/*
+ * Constants for C code to access elements in Fortran MPI status array.
+ */
+#define MPI_F_STATUS_SIZE OMPI_FORTRAN_STATUS_SIZE    /* Size of Fortran MPI status array */
+#define MPI_F_SOURCE      0                           /* Index for MPI_SOURCE */       
+#define MPI_F_TAG         1                           /* Index for MPI_TAG */
+#define MPI_F_ERROR       2                           /* Index for MPI_ERROR */
```
Comment (jsquyres) [inline ompi/include/mpi.h.in:580]: These values really should come from `ompi/include/mpif-values.pl`, not be hard-coded.

This script is invoked by `autogen.pl`:

https://github.com/open-mpi/ompi/blob/59928bb5af51dec5ee7a9f96466ef6fec991deec/autogen.pl#L1461-L1479

(it looks like there's a little left-over infrastructure from when there was more than one script, but you get the idea)

I'm not sure what the right method is to flow the values from `mpif-values.pl` to `mpi.h.in` -- you don't want to just overwrite `mpi.h.in` because:

1. That would cause the git tree to be dirty
2.  `mpi.h.in` is, itself, a template file, which is written at `configure` time (not `autogen.pl` time)

It will probably take a little bit of thought to figure out a scheme for this -- I don't have the cycles to think this through.  Can you propose something?

**Review by hppritcha — CHANGES_REQUESTED** [review]
I'd suggest enhancing mpif-values.pl to write out a small 'c' include file in addition to the current *.h's for fortran, and add a line in mpi.h.in to include this small 'c' include file.
I think OMPI_FORTRAN_STATUS_SIZE is okay as is since its defined earlier in the generated mpi.h file.

**Comment by jsquyres** [comment]
@hppritcha I'm not sure I understand.  Are you suggesting:

1. Have `mpif-values.pl` to write out (for example) `mpif-values.c`
2. Have `mpi.h` #include `mpif-values.c`
3. Install `mpif-values.c` into $(installdir) (so that it can be included by user apps who #include "mpi.h")

**Comment by jsquyres** [comment]
I had a slightly-unconventional thought overnight.  What about:

1. Put the hard-coded values of these 4 constants in `configure.ac` (or maybe `ompi_setup_mpi_fortran.m4`?)
2. Propagate the values down to `mpif-values.pl` and `mpi.h` from there.

This is a little unconventional because the goal of putting these values in `configure.ac` is not a) to test anything, or b) to make them configurable.  But rather just so that these values are at the top of the substitution tree/mechanism, so to speak.  That way, the values can flow down to the relevant files via the normal via `AC_DEFINE` or `AC_SUBST` mechanisms.

There's probably some things that still need to be figured out here -- e.g., do we have to make `mpif-values.pl.in`? That seems undesirable.  Perhaps `mpif-values.pl` can read a text config file to find those values, or somesuch...? (but be wary of VPATH builds!). TBD / up to the implementor.

**Comment by jjhursey** [comment]
I'm not following the need for the added complexity.

The new constants are meant to help a C application deal with the Fortran status structure which is defined in `mpi.h.in`:
https://github.com/open-mpi/ompi/blob/a43c46f846d98870d39e25f78c1ce09e4420effc/ompi/include/mpi.h.in#L480-L488

MPI-4, pp. 844, ln. 5:
> In C, such an f_status array can be defined with MPI_Fint f_status[ MPI_F_STATUS_SIZE]. Within this array, one can use in C the indexes MPI_F_SOURCE, MPI_F_TAG, and MPI_F_ERROR, to access the same elements as in Fortran with MPI_SOURCE, MPI_TAG and MPI_ERROR. The C indexes are 1 less than the corresponding indexes in Fortran due to the different default array start indexes in both languages.

So the new constants are only relevant to C programs, so should be part of `mpi.h.in`. Additionally, the `MPI_F_SOURCE`, `MPI_F_TAG`, `MPI_F_ERROR` are relative to their position in the structure defined in `mpi.h.in`. So I would assume that the defines are best placed as close to that structure as possible since they are essentially hard coded positional values for the fields of the structure. The other constant `MPI_F_STATUS_SIZE` is used in the `ompi_f08_status_public_t` padding and the value associated with it (`OMPI_FORTRAN_STATUS_SIZE`) is defined by configure earlier in `mpi.h.in`.

It is my understanding that `mpif-values.pl` is only for constant values that are to be used by Fortran. So these new constants would not qualify. Note that there are quite a few hard-coded, duplicated values from `mpi.h.in` and `mpif-values.pl` (e.g., `MPI_UNDEFINED`).

The only connection I see with the `mpif-values.pl` is in the last sentence from the MPI 4.0 text:
> The C indexes are 1 less than the corresponding indexes in Fortran due to the different default array start indexes in both languages.

These values are indeed part of `mpif-values.pl`
https://github.com/open-mpi/ompi/blob/a43c46f846d98870d39e25f78c1ce09e4420effc/ompi/include/mpif-values.pl#L230-L232

So the corresponding C values need to be one less than these three values.

Suggestion - if my argument is correct:
 * Keep the definitions of `MPI_F_SOURCE`, `MPI_F_TAG`, `MPI_F_ERROR`, `MPI_F_STATUS_SIZE` in `mpi.h.in` since they are intended for C programs.
 * Add a comment above the new constants regarding their connection with other files in the repo. Something like the following:

```c
/*
 * The constants MPI_F_SOURCE, MPI_F_TAG, MPI_F_ERROR, MPI_F_STATUS_SIZE
 * are defined relative to both their position in the MPI_Status structure and their
 * corresponding constant values in Fortran (i.e., MPI_SOURCE, MPI_TAG, MPI_ERROR).
 * The Fortran constants are defined in mpif-value.pl. The MPI_Status constant is defined
 * in this header file. Since these four MPI_F_ constants are meant for C programs
 * they are listed in this header file which also includes the MPI_Status structure declaration.
 */
```

**Comment by jsquyres** [comment]
@jjhursey @dwootton-ny and I just chatted on a Webex about this.

## Short version

I'll do 3 things:

1. Approve this PR as-is
2. Make another PR to amend/fix comments in `mpi.h.in`
3. File an issue to unify all integer constants between `mpi.h.in` and `mpif-values.pl`

In turn, @dwootton-ny will:

1. Get this PR merged
2. Since this is an MPI-3.1 errata item, make corresponding cherry-pick PRs for v4.0.x, v4.1.x, and v5.0.x

## More detail

Here's a little more explanation...

My concern is that I didn't want to add numerical constants that do, in fact, have corresponding values in C and Fortran, but they're just hard-coded constants in `mpi.h.in` and `mpif-values.pl`, respectively.  Instead, I'd prefer them to come from a common source (whatever that is), so that there's only a single source of truth for these integer values.  Less chance for human error that way.

@jjhursey correctly made me understand that `mpi.h.in` is *full* of integer values that have the exact same integer values in `mpif-values.pl` -- i.e., we already have this implicit linkage problem with all the other, existing values: it's not unique to these 4 new values.  Yep -- fair point.

So let's go ahead and get these 4 new values in (since it's an MPI-3.1 errata) and then have a separate issue/PR to do a proper unification of the integer constant values between `mpi.h` and the Fortran interfaces.

Per @dalcinl's comment, I'll also file another PR that adds a comment to `mpi.h.in` that says that if you edit any of the constants here in `mpi.h.in`, you should almost certainly also edit the corresponding values in `mpif-values.pl`.  I'll also clean up a few existing comments that refer to old/stale Fortran header files (we found at least one -- I'll check the others).  Might be worth adding some additional comments around status (per @dalcinl), too.

**Review by jsquyres — APPROVED** [review]

**Comment by jsquyres** [comment]
> I'd suggest enhancing mpif-values.pl to write out a small 'c' include file in addition to the current *.h's for fortran, and add a line in mpi.h.in to include this small 'c' include file. I think OMPI_FORTRAN_STATUS_SIZE is okay as is since its defined earlier in the generated mpi.h file.

@hppritcha Can you re-review this PR based on the above comments?  Your negative review is preventing this PR from being merged.

**Review by hppritcha — APPROVED** [review]

**Comment by jjhursey** [comment]
@drwootton Please cherry-pick to the v5.0.x, v4.1.x, and v4.0.x branches. Thanks!

**Comment by dalcinl** [comment]
@jsquyres @jjhursey @drwootton I'm confused. AFAICT, the `MPI_F_{SOURCE|TAG|ERROR}` constants were added in MPI-4. I don't see them mentioned in the MPI-3.1 errata. What am I missing? If I am correct, then why should this PR be cherry-picked to v4.1.x and v4.0.x?

File: ompi/include/mpi.h.in [hunk 2]
```diff
@@ -571,6 +571,14 @@ typedef MPI_Win_errhandler_function MPI_Win_errhandler_fn
 #define MPI_MAX_PSET_NAME_LEN    OPAL_MAX_PSET_NAME_LEN /* max pset name len */
 #define MPI_MAX_STRINGTAG_LEN    OPAL_MAX_STRINGTAG_LEN /* max length of string arg to comm from group funcs*/
 
+/*
+ * Constants for C code to access elements in Fortran MPI status array.
+ */
+#define MPI_F_STATUS_SIZE OMPI_FORTRAN_STATUS_SIZE    /* Size of Fortran MPI status array */
+#define MPI_F_SOURCE      0                           /* Index for MPI_SOURCE */
```
Comment (dalcinl) [inline ompi/include/mpi.h.in:578]: @jsquyres I just noticed that this PR introduced trailing whitespace in this line. Maybe you can fix it in your upcoming PR.

**Comment by jjhursey** [comment]
Good point on the MPI 3.1. They were accepted into the v3 branch after v3.1 was release as errata. Since there is not a v3.2 (yet?) it is not finalized. 
 * Original email thread: https://lists.mpi-forum.org/pipermail/mpiwg-fortran/2020-September/003584.html
 * MPI Forum Issue: https://github.com/mpi-forum/mpi-issues/issues/306
 * MPI Forum PR for v4.0: https://github.com/mpi-forum/mpi-standard/pull/270
 * MPI Forum PR for v3.x: https://github.com/mpi-forum/mpi-standard/pull/271

@drwootton  So it only needs to be cherry-picked to the v5.0.x branch.

File: ompi/include/mpi.h.in (hunk 2 above)
Comment (jsquyres) [reply ompi/include/mpi.h.in:578]: We should have a CI check for that...  Thanks for noticing; will fix in #10528.

**Comment by jsquyres** [comment]
I don't think that's right.  Errata are in force as soon as they are voted in.  So technically, the `MPI_F_*` constants are "in" MPI-3 -- even though there's no new MPI-3 doc, and it looks like this item was erroneously left out of the MPI-3 errata doc (!).

I've talked to Wesley (MPI Forum Secretary) about this: he's going to follow up with Bill (MPI Forum Editor) and/or Martin (MPI Forum Grand Chief) and/or MPI_COMM_WORLD to make sure that the MPI-3 errata doc gets updated.

Bottom line: I still think that this should get cherry picked back to the OMPI 4.0.x and 4.1.x branches. 😄

**Comment by jsquyres** [comment]
@drwootton If everyone agrees and you make PRs for v4.0.x and v4.1.x, can you please include commit d44ec62c2a5cbee04569589c2fa18a2184c05ae8 in those PRs?  That will include the whitespace/comment fixes.  Thanks!
